File: src/server/hos/calculator.ts
/**
 * Core pure HOS calculation utilities based on FMCSA CFR Part 395
 * Interstate Truck Driver's Guide to Hours of Service (April 2022)
 */

export const HOS_CONSTANTS = {
  CYCLE_LIMIT_HOURS: 70.0,
  CYCLE_DAYS: 8,
  MAX_DAILY_DRIVING_HOURS: 11.0,
  MAX_DAILY_WINDOW_HOURS: 14.0,
  MAX_CUMULATIVE_DRIVING_BEFORE_BREAK: 8.0,
  MIN_REST_BREAK_HOURS: 0.5, // 30 minutes
  MIN_DAILY_RESET_HOURS: 10.0, // 10 consecutive hours
  MIN_CYCLE_RESTART_HOURS: 34.0, // 34-hour restart
  MAX_FUELING_INTERVAL_MILES: 1000.0,
  FUELING_TRIGGER_MILES: 920.0, // Schedule fueling before reaching 1,000 mi
  FUELING_DURATION_HOURS: 0.5, // 30 minutes on-duty
  PICKUP_DURATION_HOURS: 1.0, // 1 hour on-duty
  DROPOFF_DURATION_HOURS: 1.0 // 1 hour on-duty
};

/**
 * Calculates remaining available cycle hours under the 70h/8-day rule
 */
export function calculateRemainingCycle(currentCycleUsed: number, limit = HOS_CONSTANTS.CYCLE_LIMIT_HOURS): number {
  return Math.max(0, Math.round((limit - currentCycleUsed) * 100) / 100);
}

/**
 * Calculates remaining driving hours inside the active 14-hour window
 */
export function calculateDrivingRemaining(drivingInWindow: number, limit = HOS_CONSTANTS.MAX_DAILY_DRIVING_HOURS): number {
  return Math.max(0, Math.round((limit - drivingInWindow) * 100) / 100);
}

/**
 * Calculates remaining driving window time before driving is barred
 */
export function calculateWindowRemaining(windowElapsedTime: number, limit = HOS_CONSTANTS.MAX_DAILY_WINDOW_HOURS): number {
  return Math.max(0, Math.round((limit - windowElapsedTime) * 100) / 100);
}

/**
 * Calculates driving time allowed before mandatory 30-minute consecutive break
 */
export function calculateBreakRemaining(
  cumulativeDrivingSinceBreak: number,
  limit = HOS_CONSTANTS.MAX_CUMULATIVE_DRIVING_BEFORE_BREAK
): number {
  return Math.max(0, Math.round((limit - cumulativeDrivingSinceBreak) * 100) / 100);
}

/**
 * Checks if driving is legally permitted
 */
export function canDrive(
  drivingRemaining: number,
  windowRemaining: number,
  cycleRemaining: number,
  breakRemaining: number
): boolean {
  return drivingRemaining > 0 && windowRemaining > 0 && cycleRemaining > 0 && breakRemaining > 0;
}

/**
 * Maximum driving segment duration permissible right now given all constraints
 */
export function maxAllowedDrivingDuration(
  drivingRemaining: number,
  windowRemaining: number,
  cycleRemaining: number,
  breakRemaining: number
): number {
  const min = Math.min(drivingRemaining, windowRemaining, cycleRemaining, breakRemaining);
  return Math.max(0, Math.round(min * 100) / 100);
}

/**
 * Checks if an 8-hour cumulative driving break is required
 */
export function requires30MinuteBreak(cumulativeDrivingSinceBreak: number): boolean {
  return cumulativeDrivingSinceBreak >= HOS_CONSTANTS.MAX_CUMULATIVE_DRIVING_BEFORE_BREAK;
}

/**
 * Checks if 10-hour daily rest is required
 */
export function requiresDailyReset(drivingInWindow: number, windowElapsedTime: number): boolean {
  return (
    drivingInWindow >= HOS_CONSTANTS.MAX_DAILY_DRIVING_HOURS ||
    windowElapsedTime >= HOS_CONSTANTS.MAX_DAILY_WINDOW_HOURS
  );
}

/**
 * Checks if fuel stop is needed
 */
export function requiresFuelStop(
  distanceSinceLastFuel: number,
  threshold = HOS_CONSTANTS.FUELING_TRIGGER_MILES
): boolean {
  return distanceSinceLastFuel >= threshold;
}

File: backend/services/hos/calculator.py
"""
Core pure HOS calculation utilities based on FMCSA CFR Part 395
Interstate Truck Driver's Guide to Hours of Service (April 2022)
"""

CYCLE_LIMIT_HOURS = 70.0
CYCLE_DAYS = 8
MAX_DAILY_DRIVING_HOURS = 11.0
MAX_DAILY_WINDOW_HOURS = 14.0
MAX_CUMULATIVE_DRIVING_BEFORE_BREAK = 8.0
MIN_REST_BREAK_HOURS = 0.5   # 30 minutes
MIN_DAILY_RESET_HOURS = 10.0  # 10 consecutive hours
MIN_CYCLE_RESTART_HOURS = 34.0
MAX_FUELING_INTERVAL_MILES = 1000.0
FUELING_TRIGGER_MILES = 920.0
FUELING_DURATION_HOURS = 0.5
PICKUP_DURATION_HOURS = 1.0
DROPOFF_DURATION_HOURS = 1.0


def calculate_remaining_cycle(current_cycle_used: float, limit: float = CYCLE_LIMIT_HOURS) -> float:
    return max(0.0, round(limit - current_cycle_used, 2))


def calculate_driving_remaining(driving_in_window: float, limit: float = MAX_DAILY_DRIVING_HOURS) -> float:
    return max(0.0, round(limit - driving_in_window, 2))


def calculate_window_remaining(window_elapsed_time: float, limit: float = MAX_DAILY_WINDOW_HOURS) -> float:
    return max(0.0, round(limit - window_elapsed_time, 2))


def calculate_break_remaining(cumulative_driving_since_break: float, limit: float = MAX_CUMULATIVE_DRIVING_BEFORE_BREAK) -> float:
    return max(0.0, round(limit - cumulative_driving_since_break, 2))


def can_drive(driving_rem: float, window_rem: float, cycle_rem: float, break_rem: float) -> bool:
    return driving_rem > 0 and window_rem > 0 and cycle_rem > 0 and break_rem > 0


def requires_30_minute_break(cumulative_driving_since_break: float) -> bool:
    return cumulative_driving_since_break >= MAX_CUMULATIVE_DRIVING_BEFORE_BREAK


def requires_daily_reset(driving_in_window: float, window_elapsed_time: float) -> bool:
    return (
        driving_in_window >= MAX_DAILY_DRIVING_HOURS or
        window_elapsed_time >= MAX_DAILY_WINDOW_HOURS
    )


def requires_fuel_stop(distance_since_last_fuel: float, threshold: float = FUELING_TRIGGER_MILES) -> bool:
    return distance_since_last_fuel >= threshold
